File: src/main/scala/work/chiro/game/prop/AbstractProp.scala
package work.chiro.game.prop

import work.chiro.game.GlobalConfigLoader.config
import work.chiro.game.aircraft.AbstractAircraft
import work.chiro.game.animate.AnimateContainer
import work.chiro.game.basic.PositionType.Position
import work.chiro.game.basic.{AbstractObject, Vec2Double}

abstract class AbstractProp(posInit: Position, animateContainer: AnimateContainer[Vec2Double])
  extends AbstractObject(posInit, animateContainer) {
  override def forward() = {
    super.forward()
    // 判定 x 轴出界
    if (getLocationX <= 0 || getLocationX >= config.window.playWidth) vanish()
    // 判定 y 轴出界
    if (getLocationY <= 0 || getLocationY >= config.window.playHeight) vanish()
    // if (getSpeedY > 0 && getLocationY >= config.window.playHeight) {
    //   // 向下飞行出界
    //   vanish()
    // }
  }

  def handleAircrafts(enemyAircrafts: List[AbstractAircraft]): Unit

  def create(position: Position): AbstractProp
}

File: app/src/main/java/work/chiro/game/prop/AbstractProp.java
package work.chiro.game.prop;

import work.chiro.game.aircraft.AbstractAircraft;
import work.chiro.game.animate.AnimateContainer;
import work.chiro.game.application.MusicManager;
import work.chiro.game.basic.AbstractFlyingObject;
import work.chiro.game.bullet.BaseBullet;
import work.chiro.game.utils.Utils;
import work.chiro.game.vector.Vec2;

import java.util.LinkedList;
import java.util.List;

/**
 * 道具类。
 * 也可以考虑不同类型的子弹
 *
 * @author hitsz
 */
abstract public class AbstractProp extends AbstractFlyingObject implements PropApplier {
    protected List<AbstractAircraft> enemyAircrafts = new LinkedList<>();
    protected List<BaseBullet> enemyBullets = new LinkedList<>();

    public AbstractProp(Vec2 posInit, AnimateContainer animateContainer) {
        super(posInit, animateContainer);
    }

    protected void playSupplyMusic() {
        Utils.startMusic(MusicManager.MusicType.PROPS);
    }

    public AbstractProp subscribeEnemyAircrafts(List<AbstractAircraft> enemyAircrafts) {
        this.enemyAircrafts = enemyAircrafts;
        return this;
    }

    public AbstractProp subscribeEnemyBullets(List<BaseBullet> enemyBullets) {
        this.enemyBullets = enemyBullets;
        return this;
    }
}
